### PR TITLE
Fix SYCL atomics for local memory

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/4582. @brian-kelley Can you please check if this fixes the local atomics issue for you?

### CODE REVIEW

**Comment by masterleinad** [comment]
Also, see https://github.com/intel/llvm/issues/4669#issuecomment-930398082. We would likely replace the checks here with `address_space_cast` as soon as it's available in the compiler or use `generic_space` as soon as it's available.
IMHO, passing the address space from the outside is likely neither an option simplicity (`sycl::local_ptr`, `sycl::global_ptr`) nor explicitly.

**Comment by brian-kelley** [comment]
@masterleinad Wow, that was fast. Yes, it did fix local atomics :)

**Comment by masterleinad** [comment]
Now, we just need to find a way to make that work for `SYCL+CUDA`...

**Comment by masterleinad** [comment]
I decided, for now, to only fix the behavior for Intel GPUs. I'm happy to create a corresponding pull request on the `desul` side if we are happy with this.

File: core/src/desul/atomics/SYCL.hpp [hunk 1]
```diff
@@ -17,17 +17,69 @@ SPDX-License-Identifier: (BSD-3-Clause)
...
-    Impl::sycl_atomic_ref<TYPE, MemoryOrder, MemoryScopeDevice> dest_ref(*dest);   \
+    auto l = __SYCL_GenericCastToPtrExplicit_ToLocal<TYPE>(dest);                  \
+    if (l) {                                                                       \
+      Impl::sycl_atomic_ref<TYPE,                                                  \
+                            MemoryOrder,                                           \
+                            MemoryScopeDevice,                                     \
+                            sycl::access::address_space::local_space>              \
+          dest_ref(*dest);                                                         \
+      return dest_ref.fetch_##OPER(val);                                           \
+    } else {                                                                       \
```
Comment (dalg24) [inline core/src/desul/atomics/SYCL.hpp:34]: Why is it not
```C++
auto g = __SYCL_GenericCastToPtrExplicit_ToGlobal<TYPE>(dest);
if (g) { ...
```

File: core/src/desul/atomics/SYCL.hpp [hunk 2]
```diff
@@ -17,17 +17,69 @@ SPDX-License-Identifier: (BSD-3-Clause)
 
 namespace desul {
 
+// FIXME_SYCL We need to either use generic_space or figure out how to check for the
+// correct adress space in a SYCL-portable way.
+#if defined(__SYCL_DEVICE_ONLY__) && !defined(__NVPTX__)
```
Comment (dalg24) [inline core/src/desul/atomics/SYCL.hpp:22]:
```suggestion
#ifdef __NVPTX__
```

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/desul/atomics/SYCL.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/desul/atomics/SYCL.hpp:34]: Do you want both checks? 
For one check, I choose the space to test pretty arbitrarily. Do you have any arguments for rather checking for the global space?

File: core/src/desul/atomics/SYCL.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/desul/atomics/SYCL.hpp:22]: Why? I need both checks for `__SYCL_GenericCastToPtrExplicit_ToLocal` being available.

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/desul/atomics/SYCL.hpp (hunk 2 above)
Comment (dalg24) [reply core/src/desul/atomics/SYCL.hpp:22]: https://github.com/kokkos/kokkos/blob/48beb0b7192d070030a118d6c375efc69cbe0b12/core/src/desul/atomics/Macros.hpp#L35-L37

Comment (masterleinad) [reply core/src/desul/atomics/SYCL.hpp:22]: OK. I would consider restricting the SYCL atomics to device only more like a temporary solution but I don't feel overly strong about it. Do you want me to change?

Comment (dalg24) [reply core/src/desul/atomics/SYCL.hpp]: The code is guarded by https://github.com/kokkos/kokkos/blob/2307ce3271c92d9d21307f7159eea4c6858e5c59/core/src/desul/atomics/Compare_Exchange_SYCL.hpp#L19 so this is redundant

Comment (masterleinad) [reply core/src/desul/atomics/SYCL.hpp]: I already removed it in [`2307ce3` (#4585)](https://github.com/kokkos/kokkos/pull/4585/commits/2307ce3271c92d9d21307f7159eea4c6858e5c59).

File: core/src/desul/atomics/SYCL.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/desul/atomics/SYCL.hpp:34]: No I was just wondering whether we should check that the pointer is indeed global.

**Comment by dalg24** [comment]
> I decided, for now, to only fix the behavior for Intel GPUs. I'm happy to create a corresponding pull request on the `desul` side if we are happy with this.

Please do so

**Comment by masterleinad** [comment]
> Please do so

See https://github.com/desul/desul/pull/48.

File: core/src/desul/atomics/SYCL.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/desul/atomics/SYCL.hpp:34]: I don't feel overly strong about this. As of know, there is https://github.com/intel/llvm/blob/a3e9aab6a25a9fc69e805f1a93f409375ce1f7d1/sycl/include/CL/sycl/atomic_ref.hpp#L42-L49 but also https://github.com/intel/llvm/blob/a3e9aab6a25a9fc69e805f1a93f409375ce1f7d1/sycl/include/CL/sycl/atomic_ref.hpp#L116-L120. All that is to say that really only `global_space` and `local_space` are supported (and we would get rid of all of this as soon as `generic_space` is available).
At this point, I slightly prefer not to have another check for every call SYCL atomics. If you (or others) prefer we can add the check and adopt https://github.com/intel/llvm/blob/a3e9aab6a25a9fc69e805f1a93f409375ce1f7d1/sycl/include/sycl/ext/oneapi/sub_group.hpp#L442-L443.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**